core/kernel/interrupt.c

```diff
@@ -23,27 +23,28 @@ void itr_init(struct itr_chip *chip)
 	itr_chip = chip;
 }
 
-static struct itr_handler *find_handler(size_t it)
-{
-	struct itr_handler *h;
-
-	SLIST_FOREACH(h, &handlers, link)
-		if (h->it == it)
-			return h;
-	return NULL;
-}
-
 void itr_handle(size_t it)
 {
-	struct itr_handler *h = find_handler(it);
-
-	if (!h) {
-		EMSG("Disabling unhandled interrupt %zu", it);
-		itr_chip->ops->disable(itr_chip, it);
-		return;
+	struct itr_handler *h = NULL;
+	enum itr_return ret = ITRR_NONE;
+	bool was_handled = false;
+
+	SLIST_FOREACH(h, &handlers, link) {
+		if (h->it != it)
+			continue;
+
+		ret = h->handler(h);
+
+		if (ret == ITRR_HANDLED) {
+			was_handled = true;
+		} else if (!(h->flags & ITRF_SHARED)) {
+			EMSG("Disabling interrupt %zu for not shared irq", it);
+			itr_chip->ops->disable(itr_chip, it);
+			return;
```

Comment:
could replace with a `break` here.

Reply:
I mean, I think we don't need a specific trace for shared interrupt, IMO.

Reply:
```c

	SLIST_FOREACH(h, &handlers, link) {
		if (h->it == it) {
			if (h->handler(h) == ITRR_HANDLED)
				was_handled = true;
			else if (!(h->flags & ITRF_SHARED))
				break;
		}
	}

Reply:
I see an issue for that scheme. Consider an interrupt shared by 2 consumers, which one of the 2 consumers will clear the interrupt? I would think the consumer that returned `ITRR_HANDLED` would clear the interrupt.

If so, imagine an interrupt shared by consumers A and B is raised and consumer A is first in the `handlers` list, then the sequence is:
step1- interrupt raised, `itr_handle()` is called
step2- consumer A handler is called and return `ITRR_NONE`
step3- consumer B handlers is called, it clears the interrupt and return `ITRR_HANDLED`
step4- we're done.

Now imaging the interrupt is raised again between step2 and step3, and it targets consumer A. Then consumer A handler is not called after consumer B handler which has cleared the interrupt => consumer A will never be notified of its interrupt.

Reply:
Handlers are as I understand expected to do whatever needs to be done for the interrupt "pin" of the device to return to the resting state in the case of level triggered interrupts. For edge triggered I assume that the device will do another pulse if something happens again which will trigger another interrupt even if the current one is being served. 

If something happens between step2 and step3 above another interrupt is supposed to be delivered.

The clearing of the interrupt in the interrupt controller is done by the caller of this function, currently `gic_it_handle()`.

Reply:
> 
> 
> ```c
> 	SLIST_FOREACH(h, &handlers, link) {
> 		if (h->it == it) {
> 			if (h->handler(h) == ITRR_HANDLED)
> 				was_handled = true;
> 			else if (!(h->flags & ITRF_SHARED))
> 				break;
> 		}
> 	}
> ```

Hi @etienne-lms  , 

Modified version pushed, please review , 

> Now imaging the interrupt is raised again between step2 and step3, and it targets consumer A.

Is this belong to nest interrupt? I am not sure optee-os would process this case ?